sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc

```diff
@@ -0,0 +1,131 @@
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Introduction
+This extension adds the ability for device code to call the C++ `assert()`
+macro. The behavior of `assert()` in device code is similar to its behavior in
+host code. If the asserted condition is false, a message is printed to `stderr`
+and then the program typically aborts. The mechanism for aborting the program is
+different, though. Whereas the host version calls `std::abort()`, the device
+version causes an asynchronous SYCL `exception` with the
+`errc::ext_oneapi_assert` error code to be thrown. The application can catch
+this exception like any other asynchronous exception that is thrown from an
+executing kernel. The numeric value of this enumeration element is defined by
+implementation.
+
+
+Catching the asynchronous exception with `sycl::errc::ext_oneapi_assert` error
+code means that assert failure had happened in device code during kernel
+execution and the assert message is already printed to `stderr`. Format of the
+assert message is the following:
+```
+<file>:<line>: <function>: global id: [<gid0>,<gid1>,<gid2>], local id: [<lid0>,<lid1>,<lid2>] Assertion `<expr>` failed.
+```
+
+It is unspecified whether a failing `assert()` returns to its caller before the
+kernel terminates. If a failing call returns, the device code may need to
+continue execution without deadlocking for the assertion message to be printed
+or for the exception to be thrown.
+
+The contents of the exception's `e.what()` string is unspecified. Since the
+assertion message is already printed to `stderr` by the time the exception is
+thrown, the `e.what()` string may not include the location of the failing
+assertion.
+
+The `assert()` macro is defined in system include headers, not in SYCL headers.
+On most of systems it is `<cassert>` and/or `<assert.h>` header files.
+The user can can disable assertions in device code by defining the `NDEBUG`
+preprocessor macro at the point when `<sycl.hpp>` and `<cassert>/<assert.h>`
+are included.
+
+
+== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an implementation
+supporting this extension must predefine the macro `SYCL_EXT_ONEAPI_ASSERT` to
+one of the values defined in the table below. Applications can test for the
+existence of this macro to determine if the implementation supports this
+feature, or applications can test the macro’s value to determine which of the
+extension’s APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
```

Comment:
It is customary to include an example to the extension specification that illustrates how to use the extension.  I'd recommend moving the example that you have in the design document and put it here instead.

If you feel that you also need the example in the design document, then copy the example here, so that it is in both places.

Reply:
Added.